minio/crypto.py

```diff
@@ -77,59 +117,106 @@ def encrypt(payload: bytes, password: str) -> bytes:
     return bytes(payload)
 
 
+def _split_payload(payload: bytes) -> typing.List[bytes]:
+    """Split bigger encrypted messages (payload)"""
+    max_package = _STD_PACKAGE_SIZE + _TAG_LEN
+
+    if len(payload) > max_package:
+        return [
+            payload[i : i + max_package]
+            for i in range(0, len(payload), max_package)
+        ]
+
+    return [payload]
+
+
+def _decrypt_single(
+    payloads: typing.List[bytes], idx: int, cipher_gen
+) -> bytes:
+    """Perform decryption on the single splitted payload"""
+    payload = payloads[idx]
+
+    cipher = cipher_gen(idx, idx == len(payloads) - 1)
+
+    hmac_tag = payload[-_TAG_LEN:]
+    encrypted_data = payload[:-_TAG_LEN]
+
+    decrypted_data = cipher.decrypt_and_verify(encrypted_data, hmac_tag)
+    return decrypted_data
+
+
+def _get_cipher_provider(
+    cipher_id: int,
+) -> typing.Union[AesGcmCipherProvider, ChaCha20Poly1305CipherProvider, None]:
+    """Generate cipher basing on the version byte"""
+    if cipher_id == 0:
+        return AesGcmCipherProvider()
+    if cipher_id == 1:
+        return ChaCha20Poly1305CipherProvider()
+    return None
+
+
 def decrypt(payload: bytes, password: str) -> bytes:
```

Comment:
We would need to have reader interface than handling `bytes` here.

I am working on this including size > 16KiB for encrypt and decrypt.